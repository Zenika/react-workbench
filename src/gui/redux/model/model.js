import { RECEIVED_DOCGEN } from '../docgen'
import { docgenToModel, convertFromGuiValue } from '../utils/docgen.js'

export const UPDATE_VALUE = 'UPDATE_VALUE'

export default (state = {}, { type, payload }) => {
  switch (type) {
    case RECEIVED_DOCGEN:
      return docgenToModel(payload)
    case UPDATE_VALUE: {
      return { ...state, [payload.name]: convertFromGuiValue(payload.value, payload.type) }
    }
    default:
      return state
  }
}
