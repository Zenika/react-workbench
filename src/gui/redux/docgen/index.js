export { default } from './docgen'
export * from './docgen.selectors'
export * from './docgen.api'
