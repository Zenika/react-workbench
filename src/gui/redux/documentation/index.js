export { default } from './documentation'
export * from './documentation.api'
