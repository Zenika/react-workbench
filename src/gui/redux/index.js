export * from './store'
export { default } from './store'
