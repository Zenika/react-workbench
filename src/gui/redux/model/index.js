export { default } from './model'
export * from './model.selectors'
export * from './model.actions'
