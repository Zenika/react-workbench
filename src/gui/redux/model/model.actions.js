export const SET_MODEL = 'SET_MODEL'
export const setModel = model => ({ type: SET_MODEL, payload: model })

export const SET_VALUE = 'SET_VALUE'
export const setValue = (name, value) => ({ type: SET_VALUE, payload: { name, value } })

const getDocgenValue = (value, type) => {
  switch (type) {
    case 'bool':
      return value === 'true'
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'symbol':
      return value
    default:
      return eval(value) // eslint-disable-line no-eval
  }
}

export const fetchModel = () => (dispatch) => {
  fetch('/api/docgen').then(response => response.json()).then((json) => {
    if (json && json.length > 0) {
      dispatch(
        setModel(
          Object.keys(json[0].props || {}).reduce((obj, key) => {
            const prop = json[0].props[key]
            const type = prop.type && prop.type.name
            const value = prop.defaultValue && prop.defaultValue.value

            return {
              ...obj,
              [key]: {
                value: getDocgenValue(value, type),
                type,
              },
            }
          }, {})
        )
      )
    }
  })
}
