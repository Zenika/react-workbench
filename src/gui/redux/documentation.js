import { combineReducers } from 'redux'
import { simpleObject } from 'trampss-redux-factory'

const options = { path: 'documentation' }
export const readme = simpleObject({ ...options, name: 'readme' })
export const html = simpleObject({ ...options, name: 'html' })

export default combineReducers({
  readme,
  html,
})
