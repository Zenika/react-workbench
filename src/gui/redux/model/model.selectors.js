import { createSelector } from 'reselect'

import model from './model'
import { convertToGuiValue } from '../utils/docgen'

const getModelData = state => model.get()(state)

const getModelArray = state => model.getAsArray(state)

const getComponentValue = (name, type) =>
  createSelector(getModelData, data => convertToGuiValue(data[name].value, type))

const getComponentProps = createSelector(getModelArray, array =>
  array.map(p => ({ [p.name]: p.value })).reduce((acc, curr) => ({ ...acc, ...curr }), {})
)

export { getComponentProps, getComponentValue }
