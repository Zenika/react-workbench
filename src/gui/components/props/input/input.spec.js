/* eslint-disable react/jsx-filename-extension */
/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import snap from 'misc/test/snap'
import Input from './input'
import InputContainer from './input.container'
import { setValue, SET_VALUE } from '../../../redux/model'

describe('component/props/input', () => {
  describe('input.jsx', () => {
    const snapshot = snap(Input)
    const component = { name: 'foo', type: 'text', onChange: jest.fn() }

    describe('common props', () => {
      it('should take custom className', snapshot({ ...component, className: 'custom' }))
      it('should take custom style', snapshot({ ...component, style: { backgroundColor: 'red' } }))
    })

    describe('input type="text"', () => {
      it('should render a textfield', snapshot(component))
      it('should set a default value', snapshot({ ...component, value: 'bar' }))
      it('should trigger onChange events', () => {
        // data & mocks
        const props = { ...component, onChange: jest.fn() }
        const wrapper = mount(<Input {...props} />)

        // simulate
        wrapper.find('input').simulate('change', { target: { value: 'baz' } })

        // asserts
        expect(props.onChange.mock.calls.length).toBe(1)
        expect(props.onChange.mock.calls[0][0].target.value).toBe('baz')
      })
    })

    describe('input type="checkbox"', () => {
      it('should render a checkbox', snapshot({ ...component, type: 'checkbox' }))
      it('should set a default value', snapshot({ ...component, value: true }))
      it('should trigger onChange events', () => {
        // data & mocks
        const props = { ...component, type: 'checkbox', onChange: jest.fn() }
        const wrapper = mount(<Input {...props} />)

        // simulate
        wrapper.find('input').simulate('change', { target: { checked: true } })

        // asserts
        expect(props.onChange.mock.calls.length).toBe(1)
        expect(props.onChange.mock.calls[0][0].target.checked).toBe(true)
      })
    })
  })

  describe('input.container.js', () => {
    // utility function to create a snapshot of the component
    const snapshotContainer = (name, store) => {
      const container = renderer.create(
        <Provider store={store}>
          <InputContainer name={name} />
        </Provider>
      )
      const tree = container.toJSON()
      expect(tree).toMatchSnapshot()
    }

    describe('model with type string', () => {
      it('should render a textfield', () => {
        const store = createStore(() => ({ model: { foo: { type: 'string' } } }))
        snapshotContainer('foo', store)
      })
      it('should render a textfield with value', () => {
        const store = createStore(() => ({ model: { foo: { type: 'string', value: 'bar' } } }))
        snapshotContainer('foo', store)
      })
      it(`should trigger dispatch ${SET_VALUE} when value was changed`, () => {
        const store = createStore(() => ({ model: { foo: { type: 'string', value: 'bar' } } }))
        store.dispatch = jest.fn()

        const wrapper = mount(
          <Provider store={store}>
            <InputContainer name="foo" />
          </Provider>
        )

        wrapper.find('input').simulate('change', { target: { value: 'baz' } })

        expect(store.dispatch.mock.calls.length).toBe(1)
        expect(store.dispatch.mock.calls[0]).toEqual([setValue('foo', 'baz')])
      })
    })

    describe('model with type bool', () => {
      it('should render a checkbox', () => {
        const store = createStore(() => ({ model: { foo: { type: 'bool' } } }))
        snapshotContainer('foo', store)
      })
      it('should render a checkbox with value', () => {
        const store = createStore(() => ({ model: { foo: { type: 'bool', value: true } } }))
        snapshotContainer('foo', store)
      })
      it(`should trigger dispatch ${SET_VALUE} when value was changed`, () => {
        const store = createStore(() => ({ model: { foo: { type: 'bool', value: false } } }))
        store.dispatch = jest.fn()

        const wrapper = mount(
          <Provider store={store}>
            <InputContainer name="foo" />
          </Provider>
        )

        wrapper.find('input').simulate('change', { target: { checked: true } })

        expect(store.dispatch.mock.calls.length).toBe(1)
        expect(store.dispatch.mock.calls[0]).toEqual([setValue('foo', true)])
      })
    })

    describe('model with type array', () => {
      it('should render a textfield', () => {
        const store = createStore(() => ({ model: { foo: { type: 'array' } } }))
        snapshotContainer('foo', store)
      })
      it('should render a textfield with value', () => {
        const store = createStore(() => ({ model: { foo: { type: 'array', value: ['item1', 'item2'] } } }))
        snapshotContainer('foo', store)
      })
    })

    describe('model with type object', () => {
      it('should render a textfield', () => {
        const store = createStore(() => ({ model: { foo: { type: 'object' } } }))
        snapshotContainer('foo', store)
      })
      it('should render a textfield with value', () => {
        const store = createStore(() => ({ model: { foo: { type: 'object', value: { item: 'value' } } } }))
        snapshotContainer('foo', store)
      })
    })

    describe('model with type function', () => {
      it('should render a textfield', () => {
        const store = createStore(() => ({ model: { foo: { type: 'func' } } }))
        snapshotContainer('foo', store)
      })
      it('should render a textfield with value', () => {
        const store = createStore(() => ({ model: { foo: { type: 'func', value: '() => {}' } } }))
        snapshotContainer('foo', store)
      })
    })
  })
})
