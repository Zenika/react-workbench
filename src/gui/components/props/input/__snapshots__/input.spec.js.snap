// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/props/input input.container.js model with type array should render a textfield 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      array
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.container.js model with type array should render a textfield with value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      array
    </small>
  </div>
  <input
    checked="[\\"item1\\",\\"item2\\"]"
    onChange={[Function]}
    type="text"
    value="[\\"item1\\",\\"item2\\"]"
  />
</div>
`;

exports[`component/props/input input.container.js model with type bool should render a checkbox 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      bool
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.container.js model with type bool should render a checkbox with value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      bool
    </small>
  </div>
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
    value={true}
  />
</div>
`;

exports[`component/props/input input.container.js model with type function should render a textfield 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      func
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.container.js model with type function should render a textfield with value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      func
    </small>
  </div>
  <input
    checked={[Function]}
    onChange={[Function]}
    type="text"
    value={[Function]}
  />
</div>
`;

exports[`component/props/input input.container.js model with type object should render a textfield 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      object
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.container.js model with type object should render a textfield with value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      object
    </small>
  </div>
  <input
    checked="{\\"item\\":\\"value\\"}"
    onChange={[Function]}
    type="text"
    value="{\\"item\\":\\"value\\"}"
  />
</div>
`;

exports[`component/props/input input.container.js model with type string should render a textfield 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.container.js model with type string should render a textfield with value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked="foo"
    onChange={[Function]}
    type="text"
    value="foo"
  />
</div>
`;

exports[`component/props/input input.jsx common props should take custom className 1`] = `
<div
  className="custom"
  style={Object {}}
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.jsx common props should take custom style 1`] = `
<div
  className=""
  style={
    Object {
      "backgroundColor": "red",
    }
  }
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.jsx input type="checkbox" should render a checkbox 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      bool
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.jsx input type="checkbox" should set a default value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked={true}
    onChange={[Function]}
    type="text"
    value={true}
  />
</div>
`;

exports[`component/props/input input.jsx input type="text" should render a textfield 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked={undefined}
    onChange={[Function]}
    type="text"
    value={undefined}
  />
</div>
`;

exports[`component/props/input input.jsx input type="text" should set a default value 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    foo
     
    <small>
      string
    </small>
  </div>
  <input
    checked="bar"
    onChange={[Function]}
    type="text"
    value="bar"
  />
</div>
`;
